File: filecrypt.go
// filecrypt.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"os"
	"golang.org/x/crypto/pbkdf2"
)

// ANSI colors
const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
)

func colorize(text, color string) string {
	return color + text + reset
}

const (
	saltLen   = 16
	nonceLen  = 12
	tagLen    = 16
	keyLen    = 32
	iterations = 100000
	chunkSize = 64 * 1024
)

func deriveKey(password, salt []byte) []byte {
	return pbkdf2.Key(password, salt, iterations, keyLen, sha256.New)
}

func encryptFile(inputPath, outputPath, password string) error {
	// Generate salt and nonce
	salt := make([]byte, saltLen)
	if _, err := rand.Read(salt); err != nil {
		return err
	}
	nonce := make([]byte, nonceLen)
	if _, err := rand.Read(nonce); err != nil {
		return err
	}
	key := deriveKey([]byte(password), salt)

	block, err := aes.NewCipher(key)
	if err != nil {
		return err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return err
	}

	inFile, err := os.Open(inputPath)
	if err != nil {
		return err
	}
	defer inFile.Close()

	outFile, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer outFile.Close()

	// Write salt and nonce
	if _, err := outFile.Write(salt); err != nil {
		return err
	}
	if _, err := outFile.Write(nonce); err != nil {
		return err
	}

	// Read entire file (for simplicity; for large files use streaming with Seal)
	plaintext, err := io.ReadAll(inFile)
	if err != nil {
		return err
	}

	ciphertext := gcm.Seal(nil, nonce, plaintext, nil)
	if _, err := outFile.Write(ciphertext); err != nil {
		return err
	}

	fmt.Println(colorize("File encrypted successfully: " + outputPath, green))
	return nil
}

func decryptFile(inputPath, outputPath, password string) error {
	inFile, err := os.Open(inputPath)
	if err != nil {
		return err
	}
	defer inFile.Close()

	salt := make([]byte, saltLen)
	if _, err := io.ReadFull(inFile, salt); err != nil {
		return err
	}
	nonce := make([]byte, nonceLen)
	if _, err := io.ReadFull(inFile, nonce); err != nil {
		return err
	}

	key := deriveKey([]byte(password), salt)
	block, err := aes.NewCipher(key)
	if err != nil {
		return err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return err
	}

	ciphertext, err := io.ReadAll(inFile)
	if err != nil {
		return err
	}

	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return errors.New("decryption failed (wrong password or corrupted file)")
	}

	outFile, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer outFile.Close()
	if _, err := outFile.Write(plaintext); err != nil {
		return err
	}

	fmt.Println(colorize("File decrypted successfully: " + outputPath, green))
	return nil
}

func main() {
	if len(os.Args) < 5 {
		fmt.Println(colorize("Usage: filecrypt encrypt|decrypt <input> <output> <password>", yellow))
		os.Exit(1)
	}
	mode := os.Args[1]
	input := os.Args[2]
	output := os.Args[3]
	password := os.Args[4]

	var err error
	switch mode {
	case "encrypt":
		err = encryptFile(input, output, password)
	case "decrypt":
		err = decryptFile(input, output, password)
	default:
		fmt.Println(colorize("Invalid mode. Use encrypt or decrypt.", red))
		os.Exit(1)
	}
	if err != nil {
		fmt.Println(colorize("Error: "+err.Error(), red))
		os.Exit(1)
	}
}

File: filecrypt.py
# filecrypt.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import argparse
import base64
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.backends import default_backend
import secrets

# ANSI colors
COLORS = {
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'reset': '\033[0m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

SALT_LEN = 16
NONCE_LEN = 12
TAG_LEN = 16
KEY_LEN = 32
ITERATIONS = 100000
CHUNK_SIZE = 64 * 1024  # 64 KB

def derive_key(password, salt):
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=KEY_LEN,
        salt=salt,
        iterations=ITERATIONS,
        backend=default_backend()
    )
    return kdf.derive(password.encode('utf-8'))

def encrypt_file(input_file, output_file, password):
    # Generate salt and nonce
    salt = secrets.token_bytes(SALT_LEN)
    nonce = secrets.token_bytes(NONCE_LEN)
    key = derive_key(password, salt)
    aesgcm = AESGCM(key)

    with open(input_file, 'rb') as fin, open(output_file, 'wb') as fout:
        # Write salt, nonce
        fout.write(salt)
        fout.write(nonce)

        # Encrypt and write in chunks
        while True:
            chunk = fin.read(CHUNK_SIZE)
            if not chunk:
                break
            # For GCM, we need to encrypt all at once; but we can split,
            # however GCM is not streamable. We'll encrypt whole file at once.
            # To support large files, we'd need to use incremental AEAD, but Python's AESGCM doesn't support it easily.
            # We'll read all data into memory for simplicity (but this limits file size).
            # For a real implementation, we could use a streaming mode like AES-CTR + HMAC.
            # To keep it consistent, we'll do full-file encryption.
        # Actually, let's read entire file content (if too large, may fail).
        # Better: Use a streaming approach with AES-GCM using chunks? Not trivial.
        # We'll implement simple full-load for demonstration.
    # Re-open and read all content
    with open(input_file, 'rb') as f:
        plaintext = f.read()
    ciphertext = aesgcm.encrypt(nonce, plaintext, None)
    # ciphertext includes tag at the end (16 bytes)
    with open(output_file, 'wb') as f:
        f.write(salt)
        f.write(nonce)
        f.write(ciphertext)
    print(colorize(f"File encrypted successfully: {output_file}", 'green'))

def decrypt_file(input_file, output_file, password):
    with open(input_file, 'rb') as f:
        salt = f.read(SALT_LEN)
        nonce = f.read(NONCE_LEN)
        ciphertext = f.read()  # includes tag
    if len(salt) != SALT_LEN or len(nonce) != NONCE_LEN:
        raise ValueError("Invalid file format")
    key = derive_key(password, salt)
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(nonce, ciphertext, None)
    with open(output_file, 'wb') as f:
        f.write(plaintext)
    print(colorize(f"File decrypted successfully: {output_file}", 'green'))

def main():
    parser = argparse.ArgumentParser(description="FileCrypt – шифратор файлов AES-256-GCM")
    parser.add_argument('mode', choices=['encrypt', 'decrypt'], help='Режим работы')
    parser.add_argument('input', help='Входной файл')
    parser.add_argument('output', help='Выходной файл')
    parser.add_argument('password', help='Пароль для шифрования')
    args = parser.parse_args()

    try:
        if args.mode == 'encrypt':
            encrypt_file(args.input, args.output, args.password)
        else:
            decrypt_file(args.input, args.output, args.password)
    except Exception as e:
        print(colorize(f"Ошибка: {e}", 'red'))
        sys.exit(1)

if __name__ == '__main__':
    main()
